File: Java/_144.java
// Node, Left, Right

class Solution {
    public List<Integer> preorderTraversal(TreeNode current) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> l = new ArrayList<>();

        while (current != null || !stack.isEmpty()) {
        	if (current != null) {
        		l.add(current.val);
        		if (current.right != null) stack.push(current.right);
        		current = current.left;
        	}
        	else {
        		current = stack.pop();
        	}
        }

        return l;
    }
}

File: Python3/_144.py
class Solution:
    def preorderTraversal(self, current: Optional[TreeNode]):
        l, stack = [], []
        
        while (current or stack):
            if current:
                l.append(current.val)
                if current.right:
                    stack.append(current.right)
                    
                current = current.left
                
            else:
                current = stack.pop()
                
        return l
        